### PR TITLE
[Tezos] Use forgeOperation to sign transaction

### PR DESCRIPTION
**Pull request by lvndry**
### 📝 Description

When using `tezos.contract.transfer` in `signOperation` the function was modifying the transaction and we could not make sure that the transaction that user signed was the same as the transaction the user saw on their screen. In fact `tezos.contract.transfer` has his own fee estimation function thus we could not rely on it.
In this PR we decide to use `forgeOperations` so that we make sure that the transaction is not updated after confirmation on Ledger Live.

### ❓ Context

- **JIRA or GitHub link**: 

### ✅ Checklist

Pull Requests must pass the CI and be code reviewed. Set as Draft if the PR is not ready.

- [x] `npx changeset` was attached.
- [x] **Covered by automatic tests.** 
- [x] **Impact of the changes:** 
  - ...

---

### 🧐 Checklist for the PR Reviewers

- [ ] **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- [ ] **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- [ ] **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- [ ] **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- [ ] **Any new dependencies** have been justified and documented.
- [ ] **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)

### CODE REVIEW

File: apps/ledger-live-mobile/ios/ledgerlivemobile.xcodeproj/project.pbxproj
Comment (haammar-ledger) [inline apps/ledger-live-mobile/ios/ledgerlivemobile.xcodeproj/project.pbxproj:1]: Changes are probably unneeded?

File: libs/ledger-live-common/src/families/tezos/signOperation.ts [hunk 1]
```diff
@@ -32,53 +32,81 @@ export const signOperation: SignOperationFnSignature<Transaction> = ({
             false,
             DerivationType.ED25519,
           );
+
           tezos.setProvider({ signer: ledgerSigner });
```
Comment (haammar-ledger) [inline libs/ledger-live-common/src/families/tezos/signOperation.ts:36]: Is this still necessary, now that we call `ledgerSigner.sign` directly?

**Comment by haammar-ledger** [comment]
1 failing integration test: https://github.com/LedgerHQ/ledger-live/actions/runs/7787219111/job/21234258775?pr=6088#step:7:178

Maybe worth checking, while we're at it.

File: libs/ledger-live-common/src/families/tezos/signOperation.ts (hunk 1 above)
Comment (lvndry) [reply libs/ledger-live-common/src/families/tezos/signOperation.ts]: Seems like yes: https://github.com/LedgerHQ/ledger-live/pull/6088#issuecomment-1934713512

File: libs/ledger-live-common/src/families/tezos/signOperation.ts [hunk 2]
```diff
@@ -115,7 +171,7 @@ export const signOperation: SignOperationFnSignature<Transaction> = ({
             type: "signed",
             signedOperation: {
               operation,
-              signature: opbytes,
+              signature: signature.sbytes.slice(2),
```
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/families/tezos/signOperation.ts:174]: Can we add a comment on why this is needed ?

File: libs/ledger-live-common/src/families/tezos/signOperation.ts [hunk 3]
```diff
@@ -32,53 +33,101 @@ export const signOperation: SignOperationFnSignature<Transaction> = ({
...
+              contents.push({
+                kind: OpKind.TRANSACTION,
+                amount: transaction.amount.toString(),
+                destination: transaction.recipient,
+                source: freshAddress,
+                counter: (Number(sourceData.counter) + 1 + contents.length).toString(),
+                ...transactionFees,
+              });
+
+              forgedBytes = await rpc.forgeOperations({
```
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/families/tezos/signOperation.ts:87]: Minor but we could unit test what is provided to rpc.forgeOperations for each operation type

File: libs/ledger-live-common/src/families/tezos/signOperation.ts [hunk 4]
```diff
@@ -32,53 +33,101 @@ export const signOperation: SignOperationFnSignature<Transaction> = ({
...
+            gas_limit: (transaction.gasLimit || 0).toString(),
+            storage_limit: (transaction.storageLimit || 0).toString(),
           };
 
+          const contents: OperationContents[] = [];
+
+          if (!(account as TezosAccount).tezosResources.revealed) {
+            const revealFees = await tezos.estimate.reveal();
+
+            contents.push({
```
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/families/tezos/signOperation.ts:63]: Clean solution !

**Review by hedi-edelbloute — DISMISSED** [review]

File: libs/ledger-live-common/src/families/tezos/signOperation.ts [hunk 5]
```diff
@@ -32,53 +33,101 @@ export const signOperation: SignOperationFnSignature<Transaction> = ({
...
+              contents.push({
+                kind: OpKind.DELEGATION,
                 source: freshAddress,
+                counter: (Number(sourceData.counter) + 1 + contents.length).toString(),
+                ...transactionFees,
               });
-              opbytes = res.raw.opbytes;
-              type = "UNDELEGATE";
+
+              forgedBytes = await rpc.forgeOperations({
```
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/families/tezos/signOperation.ts:124]: We could technically use rpc.forgeOperation after all switch cases as we are always providing the same parameters

**Review by hedi-edelbloute — DISMISSED** [review]

**Review by hedi-edelbloute — APPROVED** [review]
nice :ok_hand:

**Review by hedi-edelbloute — DISMISSED** [review]

**Review by hedi-edelbloute — APPROVED** [review]

**Pull request merged**